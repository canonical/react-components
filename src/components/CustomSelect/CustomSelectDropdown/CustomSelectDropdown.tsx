import React, {
  FC,
  KeyboardEvent,
  LiHTMLAttributes,
  ReactNode,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import classnames from "classnames";
import { useListener } from "hooks";
import SearchBox from "components/SearchBox";

export type CustomSelectOption = LiHTMLAttributes<HTMLLIElement> & {
  value: string;
  label: ReactNode;
  // text used for search, sort and display in toggle button
  // text must be provided if label is not a string
  text?: string;
  disabled?: boolean;
};

export type Props = {
  searchable?: "auto" | "always" | "never";
  name: string;
  options: CustomSelectOption[];
  onSelect: (value: string) => void;
  onSearch?: (value: string) => void;
  onClose: () => void;
  header?: ReactNode;
  toggleId: string;
};

const DROPDOWN_MAX_HEIGHT = 16 * 30; // 30rem with base 16px
const DROPDOWN_MARGIN = 20;

export const adjustDropdownHeightBelow = (dropdown: HTMLUListElement) => {
  const dropdownRect = dropdown.getBoundingClientRect();
  const dropdownHeight = dropdown.offsetHeight;
  const viewportHeight = window.visualViewport?.height || window.innerHeight;

  // If the dropdown is cut off at the bottom of the viewport
  // adjust the height to fit within the viewport minus fixed margin.
  // This usually becomes an issue when the dropdown is at the bottom of the viewport or screen getting smaller.
  if (dropdownRect.bottom >= viewportHeight) {
    const adjustedHeight =
      dropdownHeight - dropdownRect.bottom + viewportHeight - DROPDOWN_MARGIN;
    dropdown.style.height = `${adjustedHeight}px`;
    dropdown.style.maxHeight = `${adjustedHeight}px`;
    return;
  }

  // If the dropdown does not have overflow, the dropdown should fit its content.
  const hasOverflow = dropdown.scrollHeight > dropdown.clientHeight;
  if (!hasOverflow) {
    dropdown.style.height = "auto";
    dropdown.style.maxHeight = "";
    return;
  }

  // If the dropdown is not cut off at the bottom of the viewport
  // adjust the height of the dropdown so that its bottom edge is 20px from the bottom of the viewport
  // until the dropdown max height is reached.
  const adjustedHeight = Math.min(
    viewportHeight - dropdownRect.top - DROPDOWN_MARGIN,
    DROPDOWN_MAX_HEIGHT,
  );
  dropdown.style.height = `${adjustedHeight}px`;
  dropdown.style.maxHeight = `${adjustedHeight}px`;
};

export const adjustDropdownHeightAbove = (
  dropdown: HTMLUListElement,
  search: HTMLInputElement | null,
) => {
  // The search height is subtracted (if necessary) so that no options will be hidden behind the search input.
  const searchRect = search?.getBoundingClientRect();
  const searchHeight = searchRect?.height || 0;
  const dropdownRect = dropdown.getBoundingClientRect();

  // If the dropdown does not have overflow, do not adjust.
  const hasOverflow = dropdown.scrollHeight > dropdown.clientHeight;
  if (!hasOverflow) {
    dropdown.style.height = "auto";
    dropdown.style.maxHeight = "";
    return;
  }

  // adjust the height of the dropdown so that its top edge is 20px from the top of the viewport.
  // until the dropdown max height is reached.
  // unlike the case where the dropdown is bellow the toggle, dropdown.bottom represents the available space above the toggle always.
  // this makes the calculation simpler since we only need to work with dropdown.bottom regardless if the element is cut off or not.
  const adjustedHeight = Math.min(
    dropdownRect.bottom - searchHeight - DROPDOWN_MARGIN,
    DROPDOWN_MAX_HEIGHT,
  );
  dropdown.style.height = `${adjustedHeight}px`;
  dropdown.style.maxHeight = `${adjustedHeight}px`;
};

export const dropdownIsAbove = (dropdown: HTMLUListElement) => {
  const toggle = document.querySelector(
    ".p-custom-select__toggle",
  ) as HTMLElement;
  const dropdownRect = dropdown.getBoundingClientRect();
  const toggleRect = toggle.getBoundingClientRect();
  return toggleRect.top >= dropdownRect.bottom;
};

export const adjustDropdownHeight = (
  dropdown: HTMLUListElement | null,
  search: HTMLInputElement | null,
) => {
  if (!dropdown) {
    return;
  }

  if (dropdownIsAbove(dropdown)) {
    adjustDropdownHeightAbove(dropdown, search);
    return;
  }

  adjustDropdownHeightBelow(dropdown);
};

export const getOptionText = (option: CustomSelectOption): string => {
  if (option.text) {
    return option.text;
  }

  if (typeof option.label === "string") {
    return option.label;
  }

  throw new Error(
    "CustomSelect: options must have a string label or a text property",
  );
};

const CustomSelectDropdown: FC<Props> = ({
  searchable,
  name,
  options,
  onSelect,
  onSearch,
  onClose,
  header,
  toggleId,
}) => {
  const [search, setSearch] = useState("");
  // track highlighted option index for keyboard actions
  const [highlightedOptionIndex, setHighlightedOptionIndex] = useState(0);
  // use ref to keep a reference to all option HTML elements so we do not need to make DOM calls later for scrolling
  const optionsRef = useRef<HTMLLIElement[]>([]);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const searchRef = useRef<HTMLInputElement>(null);
  const dropdownListRef = useRef<HTMLUListElement>(null);
  const isSearchable =
    searchable !== "never" &&
    options.length > 1 &&
    (searchable === "always" || (searchable === "auto" && options.length >= 5));

  useEffect(() => {
    if (dropdownRef.current) {
      const toggle = document.getElementById(toggleId);

      // align width with wrapper toggle width
      const toggleWidth = toggle?.getBoundingClientRect()?.width ?? 0;
      dropdownRef.current.style.setProperty("min-width", `${toggleWidth}px`);
    }

    setTimeout(() => {
      if (isSearchable) {
        searchRef.current?.focus();
        return;
      }

      dropdownRef.current?.focus();
    }, 100);
  }, [isSearchable, toggleId]);

  const handleResize = () => {
    adjustDropdownHeight(dropdownListRef.current, searchRef.current);
  };

  useLayoutEffect(handleResize, []);
  useListener(window, handleResize, "resize");

  // track selected index from key board action and scroll into view if needed
  useEffect(() => {
    optionsRef.current[highlightedOptionIndex]?.scrollIntoView({
      block: "nearest",
      inline: "nearest",
    });
  }, [highlightedOptionIndex]);

  const filteredOptions = onSearch
    ? options
    : options?.filter((option) => {
        if (!search || option.disabled) return true;
        const searchText = getOptionText(option) || option.value;
        return searchText.toLowerCase().includes(search);
      });

  const getNextOptionIndex = (goingUp: boolean, prevIndex: number) => {
    const increment = goingUp ? -1 : 1;
    let currIndex = prevIndex + increment;
    // skip disabled options for key board action
    while (filteredOptions[currIndex] && filteredOptions[currIndex]?.disabled) {
      currIndex += increment;
    }

    // consider upper bound for navigating down the list
    if (increment > 0) {
      return currIndex < filteredOptions.length ? currIndex : prevIndex;
    }

    // consider lower bound for navigating up the list
    return currIndex >= 0 ? currIndex : prevIndex;
  };

  // handle keyboard actions for navigating the select dropdown
  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    const upDownKeys = ["ArrowUp", "ArrowDown"];

    // prevent default browser actions for up, down, enter and escape keys
    // also prevent any other event listeners from being called up the DOM tree
    if ([...upDownKeys, "Enter", "Escape", "Tab"].includes(event.key)) {
      event.preventDefault();
      event.nativeEvent.stopImmediatePropagation();
    }

    if (upDownKeys.includes(event.key)) {
      setHighlightedOptionIndex((prevIndex) => {
        const goingUp = event.key === "ArrowUp";
        return getNextOptionIndex(goingUp, prevIndex);
      });
    }

    if (event.key === "Enter" && filteredOptions[highlightedOptionIndex]) {
      onSelect(filteredOptions[highlightedOptionIndex].value);
    }

    if (event.key === "Escape" || event.key === "Tab") {
      onClose();
    }
  };

  const handleSearch = (value: string) => {
    setSearch(value.toLowerCase());
    // reset selected index when search text changes
    setHighlightedOptionIndex(0);
    optionsRef.current = [];

    if (onSearch) {
      onSearch(value);
    }
  };

  const handleSelect = (option: CustomSelectOption) => {
    if (option.disabled) {
      return;
    }

    onSelect(option.value);
  };

  const optionItems = filteredOptions.map((option, idx) => {
    return (
      <li
        key={`${option.value}-${idx}`}
        onClick={() => handleSelect(option)}
        className={classnames(
          "p-list__item",
          "p-custom-select__option",
          "u-truncate",
          {
            disabled: option.disabled,
            highlight: idx === highlightedOptionIndex && !option.disabled,
          },
        )}
        // adding option elements to a ref array makes it easier to scroll the element later
        // else we'd have to make a DOM call to find the element based on some identifier
        ref={(el) => {
          if (!el) return;
          optionsRef.current[idx] = el;
        }}
        role="option"
        onMouseMove={() => setHighlightedOptionIndex(idx)}
      >
        <span
          className={classnames({
            "u-text--muted": option.disabled,
          })}
        >
          {option.label}
        </span>
      </li>
    );
  });

  return (
    <div
      className="p-custom-select__dropdown u-no-padding"
      role="combobox"
      onKeyDownCapture={handleKeyDown}
      // allow focus on the dropdown so that keyboard actions can be captured
      tabIndex={-1}
      ref={dropdownRef}
      onMouseDown={(e) => {
        // when custom select is used in a modal, which is a portal, a dropdown click
        // should not close the modal itself, so we stop the event right here.
        e.stopPropagation();
      }}
    >
      {isSearchable && (
        <div className="p-custom-select__search u-no-padding--bottom">
          <SearchBox
            ref={searchRef}
            id={`select-search-${name}`}
            name={`select-search-${name}`}
            type="text"
            aria-label={`Search for ${name}`}
            className="u-no-margin--bottom"
            onChange={handleSearch}
            value={search}
            autocomplete="off"
          />
        </div>
      )}
      {header}
      <ul
        className="p-list u-no-margin--bottom"
        role="listbox"
        ref={dropdownListRef}
      >
        {optionItems}
      </ul>
    </div>
  );
};

export default CustomSelectDropdown;
