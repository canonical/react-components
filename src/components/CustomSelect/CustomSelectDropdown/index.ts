export { default, getOptionText } from "./CustomSelectDropdown";
export type {
  CustomSelectOption,
  Props as CustomSelectDropdownProps,
} from "./CustomSelectDropdown";
