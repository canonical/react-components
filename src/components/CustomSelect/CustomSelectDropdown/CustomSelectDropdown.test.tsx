import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import CustomSelectDropdown, {
  CustomSelectOption,
  getNearestParentsZIndex,
} from "./CustomSelectDropdown";
import { CustomSelectDropdownProps } from ".";

// Global mocks
window.HTMLElement.prototype.scrollIntoView = jest.fn();

const defaultOptions: CustomSelectOption[] = [
  { value: "1", label: "Option 1" },
  { value: "2", label: "Option 2" },
  { value: "3", label: "Option 3" },
  { value: "4", label: "Option 4" },
  { value: "5", label: "Option 5" },
];

const defaultProps: CustomSelectDropdownProps = {
  name: "test-dropdown",
  toggleId: "test-toggle",
  onSelect: jest.fn(),
  onClose: jest.fn(),
  searchable: "auto",
  options: defaultOptions,
};

// mock the toggle for the dropdown menu
const mockToggle = (name: string, fn: () => unknown) => {
  const toggle = document.createElement("span");
  toggle.setAttribute("id", "test-toggle");
  toggle.classList.add("p-custom-select__toggle");
  document.body.appendChild(toggle);
  Object.defineProperty(toggle, name, {
    value: jest.fn(fn),
  });
  return toggle;
};

const cleanUpDOM = () => {
  const toggle = document.getElementById("test-toggle");
  if (toggle) {
    document.body.removeChild(toggle);
  }
};

describe("CustomSelectDropdown", () => {
  beforeEach(() => {
    cleanUpDOM();
    jest.resetAllMocks();
  });

  it("renders", () => {
    const options: CustomSelectOption[] = [
      { value: "1", label: "Option 1" },
      { value: "2", label: "Option 2" },
      { value: "3", label: "Option 3" },
      { value: "4", label: "Option 4" },
      { value: "5", label: "Option 5" },
    ];

    mockToggle("getBoundingClientRect", () => ({}));

    const { container } = render(
      <CustomSelectDropdown {...defaultProps} options={options} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render the search box when `searchable` is 'always'", () => {
    const options: CustomSelectOption[] = [
      { value: "1", label: "Option 1" },
      { value: "2", label: "Option 2" },
      { value: "3", label: "Option 3" },
    ];

    mockToggle("getBoundingClientRect", () => ({
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    }));

    render(
      <CustomSelectDropdown
        {...defaultProps}
        options={options}
        searchable="always"
      />,
    );

    const searchBox = screen.queryByRole("textbox", {
      name: /search for test-dropdown/i,
    });
    expect(searchBox).toBeInTheDocument();
  });

  describe("search functionality", () => {
    const setup = (
      searchable: CustomSelectDropdownProps["searchable"],
      options?: CustomSelectOption[],
    ) => {
      mockToggle("getBoundingClientRect", () => ({
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
      }));

      render(
        <CustomSelectDropdown
          {...defaultProps}
          options={options || defaultOptions}
          searchable={searchable}
        />,
      );
    };

    it("should render the search box when `searchable` is 'always'", () => {
      const options: CustomSelectOption[] = [
        { value: "1", label: "Option 1" },
        { value: "2", label: "Option 2", disabled: true },
        { value: "3", label: "Option 3" },
      ];

      setup("always", options);

      const searchBox = screen.queryByRole("textbox", {
        name: /search for test-dropdown/i,
      });
      expect(searchBox).toBeInTheDocument();
    });

    it("should render the search box when `searchable` is `auto` and there are 5 or more options", () => {
      setup("auto");

      const searchBox = screen.queryByRole("textbox", {
        name: /search for test-dropdown/i,
      });
      expect(searchBox).toBeInTheDocument();
    });

    it("should not render search box when `searchable` is `never`", () => {
      setup("never");

      const searchBox = screen.queryByRole("textbox", {
        name: /search for test-dropdown/i,
      });
      expect(searchBox).not.toBeInTheDocument();
    });

    it("should filter options based on the search input", () => {
      setup("auto");

      const searchBox = screen.queryByRole("textbox", {
        name: /search for test-dropdown/i,
      });

      const allOptions = [
        "Option 1",
        "Option 2",
        "Option 3",
        "Option 4",
        "Option 5",
      ];

      for (const option of allOptions) {
        expect(screen.getByText(option)).toBeInTheDocument();
      }

      fireEvent.change(searchBox, { target: { value: "Option 1" } });

      expect(screen.getByText("Option 1")).toBeInTheDocument();
      for (const option of allOptions.slice(1)) {
        expect(screen.queryByText(option)).not.toBeInTheDocument();
      }
    });

    it("should reset the selected index when search input changes", () => {
      setup("auto");

      const searchBox = screen.queryByRole("textbox", {
        name: /search for test-dropdown/i,
      });

      fireEvent.change(searchBox, { target: { value: "Option 2" } });

      expect(screen.getByText("Option 2")).toBeInTheDocument();
      expect(screen.getByText("Option 2").parentElement).toHaveClass(
        "highlight",
      );

      fireEvent.change(searchBox, { target: { value: "Option 1" } });

      expect(screen.getByText("Option 1")).toBeInTheDocument();
      expect(screen.getByText("Option 1").parentElement).toHaveClass(
        "highlight",
      );
    });
  });

  describe("dropdown positioning", () => {
    it("should align the dropdown width with the toggle button width", () => {
      mockToggle("getBoundingClientRect", () => ({
        width: 200,
      }));

      render(<CustomSelectDropdown {...defaultProps} />);

      const dropdown = screen.queryByRole("combobox");
      expect(dropdown).toHaveStyle(`min-width: ${200}px`);
    });
  });

  describe("keyboard navigation", () => {
    it("allows keyboard navigation with ArrowUp and ArrowDown keys", () => {
      mockToggle("", () => ({}));

      render(<CustomSelectDropdown {...defaultProps} />);

      const dropdown = screen.queryByRole("combobox");

      fireEvent.keyDown(dropdown, { key: "ArrowDown" });
      fireEvent.keyDown(dropdown, { key: "ArrowDown" });

      expect(screen.getByText("Option 3").parentElement).toHaveClass(
        "highlight",
      );

      fireEvent.keyDown(dropdown, { key: "ArrowUp" });
      fireEvent.keyDown(dropdown, { key: "ArrowUp" });

      expect(screen.getByText("Option 1").parentElement).toHaveClass(
        "highlight",
      );
    });

    it("skips disabled options during navigation", () => {
      mockToggle("", () => ({}));

      const disabledOptions: CustomSelectOption[] = [
        { value: "1", label: "Option 1" },
        { value: "2", label: "Option 2", disabled: true },
        { value: "3", label: "Option 3" },
      ];

      render(
        <CustomSelectDropdown {...defaultProps} options={disabledOptions} />,
      );

      const dropdown = screen.queryByRole("combobox");
      fireEvent.keyDown(dropdown, { key: "ArrowDown" });

      expect(screen.getByText("Option 3").parentElement).toHaveClass(
        "highlight",
      );
    });

    it("selects an option with Enter key", () => {
      mockToggle("", () => ({}));

      const onSelectMock = jest.fn();
      render(
        <CustomSelectDropdown {...defaultProps} onSelect={onSelectMock} />,
      );

      const dropdown = screen.queryByRole("combobox");

      fireEvent.keyDown(dropdown, { key: "ArrowDown" });
      fireEvent.keyDown(dropdown, { key: "ArrowDown" });
      fireEvent.keyDown(dropdown, { key: "Enter" });

      expect(onSelectMock).toHaveBeenCalledWith("3");
    });

    it("closes the dropdown with Escape or Tab key", () => {
      mockToggle("", () => ({}));

      const onCloseMock = jest.fn();
      render(<CustomSelectDropdown {...defaultProps} onClose={onCloseMock} />);

      const dropdown = screen.queryByRole("combobox");
      fireEvent.keyDown(dropdown, { key: "Escape" });
      fireEvent.keyDown(dropdown, { key: "Tab" });
      expect(onCloseMock).toHaveBeenCalledTimes(2);
    });
  });

  describe("getNearestParentsZIndex", () => {
    it("should return '0' if the element is not available", () => {
      const actual = getNearestParentsZIndex(null);
      expect(actual).toBe("0");
    });

    it("should return the element's z-index if it does not have a parent", () => {
      const div = document.createElement("div");
      div.style.zIndex = "5";
      document.body.appendChild(div);

      const actual = getNearestParentsZIndex(div);
      expect(actual).toBe("5");

      document.body.removeChild(div);
    });

    it("should return the z-index of the closest parent with a defined value", () => {
      const grandparent = document.createElement("div");
      grandparent.style.zIndex = "5";
      const parent = document.createElement("div");
      parent.style.zIndex = "auto";
      const child = document.createElement("div");
      child.style.zIndex = "auto";
      parent.appendChild(child);
      grandparent.appendChild(parent);
      document.body.appendChild(grandparent);

      const actual = getNearestParentsZIndex(child);
      expect(actual).toBe("5");

      document.body.removeChild(grandparent);
    });
  });
});
