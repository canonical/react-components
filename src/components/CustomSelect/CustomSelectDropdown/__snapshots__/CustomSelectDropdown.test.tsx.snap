// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSelectDropdown renders 1`] = `
<div>
  <div
    class="p-custom-select__dropdown u-no-padding"
    role="combobox"
    style="min-width: 0px;"
    tabindex="-1"
  >
    <div
      class="p-custom-select__search u-no-padding--bottom"
    >
      <div
        class="p-search-box u-no-margin--bottom"
      >
        <label
          class="u-off-screen"
          for="search"
        >
          Search
        </label>
        <input
          aria-label="Search for test-dropdown"
          autocomplete="off"
          class="p-search-box__input"
          id="select-search-test-dropdown"
          name="select-search-test-dropdown"
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          class="p-search-box__button"
        >
          <i
            class="p-icon--search"
          >
            Search
          </i>
        </button>
      </div>
    </div>
    <ul
      class="p-list u-no-margin--bottom"
      role="listbox"
      style="height: auto;"
    >
      <li
        class="p-list__item p-custom-select__option u-truncate highlight"
        role="option"
      >
        <span
          class=""
        >
          Option 1
        </span>
      </li>
      <li
        class="p-list__item p-custom-select__option u-truncate"
        role="option"
      >
        <span
          class=""
        >
          Option 2
        </span>
      </li>
      <li
        class="p-list__item p-custom-select__option u-truncate"
        role="option"
      >
        <span
          class=""
        >
          Option 3
        </span>
      </li>
      <li
        class="p-list__item p-custom-select__option u-truncate"
        role="option"
      >
        <span
          class=""
        >
          Option 4
        </span>
      </li>
      <li
        class="p-list__item p-custom-select__option u-truncate"
        role="option"
      >
        <span
          class=""
        >
          Option 5
        </span>
      </li>
    </ul>
  </div>
</div>
`;
